mobile/native/src/db/mod.rs

```diff
@@ -0,0 +1,58 @@
+pub mod models;
+
+use crate::db::models::LastLogin;
+use anyhow::anyhow;
+use anyhow::Context;
+use anyhow::Result;
+use diesel::r2d2;
+use diesel::r2d2::ConnectionManager;
+use diesel::r2d2::Pool;
+use diesel::r2d2::PooledConnection;
+use diesel::SqliteConnection;
+use diesel_migrations::embed_migrations;
+use diesel_migrations::EmbeddedMigrations;
+use diesel_migrations::MigrationHarness;
+use state::Storage;
+use std::sync::Arc;
+use time::OffsetDateTime;
+
+pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!();
+
+/// Wallet has to be managed by Rust as generics are not support by frb
+static DB: Storage<Arc<Pool<ConnectionManager<SqliteConnection>>>> = Storage::new();
+
+pub fn init_db(database_url: String) -> Result<()> {
+    let manager = ConnectionManager::<SqliteConnection>::new(database_url);
+    let pool = r2d2::Pool::builder()
+        .build(manager)
+        .expect("Failed to create pool.");
+
+    let mut connection = pool.get()?;
+
+    connection
+        .run_pending_migrations(MIGRATIONS)
+        .map_err(|e| anyhow!("could not run db migration: {e:#}"))?;
+    tracing::trace!("Database migration run - db initialized");
+
+    DB.set(Arc::new(pool));
+
+    Ok(())
+}
+
+fn get_db() -> Result<Arc<Pool<ConnectionManager<SqliteConnection>>>> {
+    DB.try_get().context("DB uninitialised").cloned()
+}
+
+pub fn connection() -> Result<PooledConnection<ConnectionManager<SqliteConnection>>> {
+    let pool = get_db()?;
+
+    pool.get()
+        .map_err(|e| anyhow!("cannot acquire database connection: {e:#}"))
+}
+
+pub fn update_last_login() -> Result<LastLogin> {
+    let mut db = connection()?;
+    let now = OffsetDateTime::now_utc();
+    let last_login = LastLogin::update_last_login(now, &mut db)?;
```

Comment:
nit: the last two lines could be combined to one.

```
LastLogin::update_last_login(now, &mut db)
```

Reply:
You're right, although historically we've preferred to end functions that return `Result`s with `Ok(value)`.

Reply:
Unfortunately it does not work because `LastLogin::update_last_login(now, &mut db)` returns a `diesel::result::Error`.